### PR TITLE
Sanitize filenames when creating them from character name

### PR DESCRIPTION
**Pull request by alanshutko**
Resolves CODE-3501, which was caused by using characters from a character name which are not allowed in filenames.

### CODE REVIEW

**Review by LegacyKing — COMMENTED** [review]
Looks good to me.

File: code/src/java/pcgen/gui2/PCGenFrame.java [hunk 1]
```diff
@@ -939,7 +940,8 @@ boolean showSaveCharacterChooser(CharacterFacade character)
 		File prevFile = file;
 		if (file == null || StringUtils.isEmpty(file.getName()))
 		{
-			file = new File(parentPath, character.getNameRef().get() + Constants.EXTENSION_CHARACTER_FILE);
+			String characterName = FileHelper.sanitizeFilename(character.getNameRef().get());
+			file = new File(parentPath, characterName + Constants.EXTENSION_CHARACTER_FILE);
```
Comment (javydreamercsw) [inline code/src/java/pcgen/gui2/PCGenFrame.java:944]: Any reason not to do this on the same line?

File: code/src/java/pcgen/util/FileHelper.java [hunk 2]
```diff
@@ -117,4 +117,22 @@ private static String stripOffRoot(File relative)
...
+	 *     <li>&lt;, &gt; :, ", \\, |, ?, * because they are
+	 *     <a href="https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions">restricted on Windows</a></li>
+	 *     <li>/ and U+0000 (Unicode NULL) , restricted on Linux, Mac, and Windows</li>
+	 * </ul>
+	 *
+	 * @param invalidCharacterFilename Filename to sanitize
+	 * @return string with invalid characters removed
+	 */
+	public static String sanitizeFilename(String invalidCharacterFilename) {
+		return invalidCharacterFilename.replaceAll("[\u0000\\\\/:*?\"<>|]", "");
```
Comment (javydreamercsw) [inline code/src/java/pcgen/util/FileHelper.java:136]: Null check for the parameter? Recently I started using Lombok NonNull annotation for parameters.

File: code/src/java/plugin/pcgtracker/PCGTrackerPlugin.java [hunk 3]
```diff
@@ -328,8 +329,9 @@ private boolean savePC(PlayerCharacter aPC, boolean saveas)
 
 		if (aPCFileName.isEmpty())
 		{
+			String characterName = FileHelper.sanitizeFilename(aPC.getDisplay().getDisplayName());
 			prevFile = new File(PCGenSettings.getPcgDir(),
-				aPC.getDisplay().getDisplayName() + Constants.EXTENSION_CHARACTER_FILE);
+				characterName + Constants.EXTENSION_CHARACTER_FILE);
```
Comment (javydreamercsw) [inline code/src/java/plugin/pcgtracker/PCGTrackerPlugin.java:334]: Any reason not to do this on the same line?

**Review by javydreamercsw — APPROVED** [review]

File: code/src/java/pcgen/gui2/PCGenFrame.java (hunk 1 above)
Comment (alanshutko) [reply code/src/java/pcgen/gui2/PCGenFrame.java:944]: I could do it on the same line. It seemed like the line was getting pretty long, though, and breaking it up looked a bit more clear.

File: code/src/java/pcgen/util/FileHelper.java (hunk 2 above)
Comment (alanshutko) [reply code/src/java/pcgen/util/FileHelper.java:136]: Good point. Will adjust.

Comment (alanshutko) [reply code/src/java/pcgen/util/FileHelper.java:136]: I'm checking other parts of the code to see if there is a pattern to parameter checking. Does anyone know the preferred pattern?

We could add lombok, but I'm not sure I want to add a dependency in my first PR. 

Another option would be to do an explicit check and throw a null pointer exception with clear messaging (similar to what lombok would do with @NotNull.

Option 3 would be to check for null, send something to the system logs, but return an empty string as a safe value for downstream code.

**Comment by LegacyKing** [comment]
Error is not code related. Working on cleaning up these errors.

**Comment by LegacyKing** [comment]
If this is approved, we can merge it in. The CI is failing for data not code.

**Comment by LegacyKing** [comment]
I do have a question. Why is this pointed at the 6.08 branch instead of the MASTER branch @alanshutko ?

**Comment by alanshutko** [comment]
When I last checked, 6.09/MASTER does not prefill a filename at all.

**Comment by LegacyKing** [comment]
Ah, that makes sense. I'm also disappointed we lost that feature in 6.09.

**Pull request merged**